tools/bridge/bridge/contract_validator.py

```diff
@@ -0,0 +1,31 @@
+from web3 import Web3
+from web3.contract import Contract
+from web3._utils.abi import abi_to_signature
+
+
+def validate_contract(contract: Contract) -> None:
+    """Verifies if a given contract exists on the chain.
+
+    It checks if any code is stored for the address.
+    For each entry in the ABI it will be checked if the signature hash can be
+    found within the code.
+    Throws an exception if the contract could not be verified. Else return
+    nothing.
+    """
+
+    contract_code = contract.web3.eth.getCode(contract.address)
+
+    if not contract_code:
+        raise ValueError(
+            f"The given contract address {contract.address} does not point to a contract!"
+        )
+
+    for entry in contract.abi:
+        signature = abi_to_signature(entry)
+        signature_hash = Web3.keccak(text=signature)
+
+        if signature_hash not in contract_code:
```

Comment:
I'm surprised that this works: As far as I know, for function calls only the first four bytes of the hash are used, so storing the whole hash seems unnecessary. For events I don't see a reason at all why the hash would appear in the code. Do you happen to know why this works or was it just guessing? Can this change with a new (minor) Solidity version?

It's a very helpful check though, so I'd keep it in any case.

Reply:
Good point. The tests where not exhaustive enough. I talked to Yoichi and he was unsure. As result [events are included fully]( https://solidity.readthedocs.io/en/develop/contracts.html?highlight=topic#low-level-interface-to-logs), while [functions only the first four bytes](https://solidity.readthedocs.io/en/v0.5.3/abi-spec.html#function-selector). So I would suggest to make a condition and for functions only check the first 4 bytes. Do you agree?

Reply:
I used the [Solidity ABI JSON documentation](https://solidity.readthedocs.io/en/develop/abi-spec.html#json) to make the validation stable. Moreover I switched the tests to use the home bridge contract, since we have more various description types in the internal ABI there.

Reply:
:heavy_check_mark: